src/AuthProvider.js

```diff
@@ -0,0 +1,66 @@
+import React, {
+  useState,
+  useContext,
+  createContext,
+} from 'react';
+import PropTypes from 'prop-types';
+import { useQuery, useMutation } from '@apollo/client';
+import QUERY_USERS from './graphql/queries';
+import REGISTER from './graphql/mutations';
+
+const AuthContext = createContext();
+
+export const useAuth = () => useContext(AuthContext);
+
+function useProvideAuth() {
+  const [user, setUser] = useState(localStorage.getItem('user') || undefined);
+  const { refetch: refetchUsers } = useQuery(QUERY_USERS);
+  const [addUser] = useMutation(REGISTER);
+
+  const signIn = async (username, password) => {
+    const result = await refetchUsers();
+    if (!result) throw Error('Server did not respond.');
+    const match = result.data.users.find((u) => u.username === username);
+    if (!match || match.password !== password) throw Error('Incorrect username or password.');
```

Comment:
You could move that condition to the previous line:
`result.data.users.find((u) => u.username === username && u.password === password);`